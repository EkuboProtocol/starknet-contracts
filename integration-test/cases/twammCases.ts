const THIRTY_BIPS_FEE = (2n ** 128n * 30n) / 10000n;
const ONE_PERCENT_FEE = (1n << 128n) / 100n;

export const TWAMM_POOL_CASES: Array<{
  name: string;
  pool: {
    startingTick: bigint;
    fee: bigint;
  };
  positions_liquidities: bigint[];
}> = [
  // {
  //   name: "0 liquidity, starting price of 1, 1% fee",
  //   pool: {
  //     startingTick: 0n,
  //     fee: ONE_PERCENT_FEE,
  //   },
  //   positions_liquidities: [],
  // },
  // {
  //   name: "1e18 liquidity, starting price of 1, 1% fee",
  //   pool: {
  //     startingTick: 0n,
  //     fee: ONE_PERCENT_FEE,
  //   },
  //   positions_liquidities: [10n ** 18n],
  // },
  {
    name: "1e18 liquidity, starting price of 1, 0.3% fee",
    pool: {
      startingTick: 0n,
      fee: THIRTY_BIPS_FEE,
    },
    positions_liquidities: [10n ** 18n],
  },
  // {
  //   name: "1e36 liquidity across 2 positions, starting price of 1, 1% fee",
  //   pool: {
  //     startingTick: 0n,
  //     fee: ONE_PERCENT_FEE,
  //   },
  //   positions_liquidities: [5n * 10n ** 35n, 5n * 10n ** 35n],
  // },
];

export const TWAMM_SWAP_CASES: Array<{
  name: string;
  orders: {
    relativeTimes: { start: number; end: number };
    isToken1: boolean;
    amount: bigint;
  }[];
  snapshotTimes: number[];
}> = [
  {
    name: "no swap, time passes",
    orders: [],
    snapshotTimes: [0, 1],
  },
  {
    name: "selling 1e18 token0 per second for one period",
    orders: [
      {
        relativeTimes: { start: 0, end: 16 },
        isToken1: false,
        amount: 16n * 10n ** 18n,
      },
    ],
    snapshotTimes: [0, 8, 16],
  },
  {
    name: "selling 1e18 token1 per second for one period",
    orders: [
      {
        relativeTimes: { start: 0, end: 16 },
        isToken1: true,
        amount: 16n * 10n ** 18n,
      },
    ],
    snapshotTimes: [0, 8, 16],
  },
];
