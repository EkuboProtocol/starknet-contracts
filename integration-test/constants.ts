export const MAX_TICK = 88722883n;
export const MIN_TICK = -88722883n;
export const MAX_U128 = 2n ** 128n - 1n;
