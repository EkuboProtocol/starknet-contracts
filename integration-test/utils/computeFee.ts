export function computeFee(x: bigint, fee: bigint): bigint {
  const p = x * fee;
  return p / 2n ** 128n + (p % 2n ** 128n !== 0n ? 1n : 0n);
}
