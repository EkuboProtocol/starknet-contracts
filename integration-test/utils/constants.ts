export const MAX_TICK = 88722883n;
export const MIN_TICK = -88722883n;

export const MAX_TICK_SPACING = 354892n;

export const MIN_SQRT_RATIO = 18446748437148339061n;
export const MAX_SQRT_RATIO =
  6277100250585753475930931601400621808602321654880405518632n;

export const MAX_U128 = 2n ** 128n - 1n;
