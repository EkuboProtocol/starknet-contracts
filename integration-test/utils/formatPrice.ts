import Decimal from "decimal.js-light";

export function formatPrice(sqrtRatioLimit: bigint) {
  return new Decimal(sqrtRatioLimit.toString())
    .div(new Decimal(2).pow(128))
    .pow(2)
    .toSignificantDigits(6)
    .toString();
}
