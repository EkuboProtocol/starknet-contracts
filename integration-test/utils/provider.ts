import { Account, CallData, ec, hash, RpcProvider, stark } from "starknet";

const DEVNET_URL = "http://127.0.0.1:5050";
export const provider: RpcProvider = new RpcProvider({
  nodeUrl: DEVNET_URL,
});

export async function setDevnetTime(time: number) {
  const response = await fetch(`${DEVNET_URL}/set_time`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: `{"time":${time}}`,
  });

  if (!response.ok) {
    throw new Error(`Failed to set time: ${await response.text()}`);
  }
}

const PREDECLARED_OZ_ACCOUNT_CLASS_HASH =
  "0x61dac032f228abef9c6626f995015233097ae253a7f72d68552db02f2971b8f";

export async function createAccount(): Promise<Account> {
  const privateKey = stark.randomAddress();
  const starkKeyPub = ec.starkCurve.getStarkKey(privateKey);

  // Calculate future address of the account
  const constructorCalldata = CallData.compile({
    publicKey: starkKeyPub,
  });

  const expectedAccountAddress = hash.calculateContractAddressFromHash(
    starkKeyPub,
    PREDECLARED_OZ_ACCOUNT_CLASS_HASH,
    constructorCalldata,
    0
  );

  const account = new Account(provider, expectedAccountAddress, privateKey);

  const response = await fetch(`${DEVNET_URL}/mint`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: `{"address":"${account.address}","amount":10000000000000000000000000000,"unit":"WEI","lite":true}`,
  });

  if (!response.ok) {
    throw new Error(`Failed to mint: ${await response.text()}`);
  }

  const { transaction_hash, contract_address } = await account.deployAccount(
    {
      classHash: PREDECLARED_OZ_ACCOUNT_CLASS_HASH,
      constructorCalldata,
      addressSalt: starkKeyPub,
    },
    { maxFee: 1_000_000_000_000_000_000n, nonce: 1n }
  );

  await provider.waitForTransaction(transaction_hash, { retryInterval: 0 });

  return account;
}
