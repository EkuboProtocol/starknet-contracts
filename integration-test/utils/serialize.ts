export type i129 = { mag: bigint; sign: boolean };

export function toI129(x: bigint): i129 {
  return {
    mag: x < 0n ? x * -1n : x,
    sign: x < 0n,
  };
}

export function fromI129(x: { mag: bigint; sign: boolean }): bigint {
  return x.sign ? x.mag * -1n : x.mag;
}
