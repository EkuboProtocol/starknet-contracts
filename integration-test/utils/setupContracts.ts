import { Account, Contract, num, shortString } from "starknet";
import MockERC20 from "../../target/dev/ekubo_MockERC20.contract_class.json";
import MockERC20Contract from "../../target/dev/ekubo_MockERC20.contract_class.json";
import MockERC20CASM from "../../target/dev/ekubo_MockERC20.compiled_contract_class.json";
import CoreCompiledContract from "../../target/dev/ekubo_Core.contract_class.json";
import CoreContract from "../../target/dev/ekubo_Core.contract_class.json";
import CoreCompiledContractCASM from "../../target/dev/ekubo_Core.compiled_contract_class.json";
import OwnedNFTContract from "../../target/dev/ekubo_OwnedNFT.contract_class.json";
import OwnedNFTContractCASM from "../../target/dev/ekubo_OwnedNFT.compiled_contract_class.json";
import PositionsCompiledContract from "../../target/dev/ekubo_Positions.contract_class.json";
import PositionsContract from "../../target/dev/ekubo_Positions.contract_class.json";
import PositionsCompiledContractCASM from "../../target/dev/ekubo_Positions.compiled_contract_class.json";
import TWAMMCompiledContract from "../../target/dev/ekubo_TWAMM.contract_class.json";
import TWAMMCompiledContractCASM from "../../target/dev/ekubo_TWAMM.compiled_contract_class.json";
import Router from "../../target/dev/ekubo_Router.contract_class.json";
import RouterContract from "../../target/dev/ekubo_Router.contract_class.json";
import RouterCASM from "../../target/dev/ekubo_Router.compiled_contract_class.json";
import { createAccount, provider } from "./provider";
import { getNextTransactionSettingsFunction } from "./getNextTransactionSettingsFunction";
import { deployTokens } from "./deployTokens";

export async function setupContracts(expected?: {
  core: string;
  positions: string;
  router: string;
  nft: string;
  twamm: string;
  tokenClassHash: string;
}) {
  if (expected) {
    try {
      const [ch0, ch1, ch2, ch3, ch4, c] = await Promise.all([
        provider.getClassHashAt(expected.core),
        provider.getClassHashAt(expected.positions),
        provider.getClassHashAt(expected.router),
        provider.getClassHashAt(expected.nft),
        provider.getClassHashAt(expected.twamm),
        provider.getClass(expected.tokenClassHash),
      ]);
      if (ch0 && ch1 && ch2 && ch3 && ch4 && c) return expected;
    } catch (error) {}
  }

  // starknet-devnet-rs with seed 0
  const deployer = new Account(
    provider,
    "0x64b48806902a367c8598f4f95c305e8c1a1acba5f082d294a43793113115691",
    "0x71d7bb07b9a64f6f78ac4c816aff4da9"
  );

  const simpleTokenContractDeclare = await deployer.declareIfNot({
    contract: MockERC20 as any,
    casm: MockERC20CASM as any,
  });
  const coreContractDeclare = await deployer.declareIfNot({
    contract: CoreCompiledContract as any,
    casm: CoreCompiledContractCASM as any,
  });
  const declareNftResponse = await deployer.declareIfNot({
    contract: OwnedNFTContract as any,
    casm: OwnedNFTContractCASM as any,
  });
  const positionsDeclare = await deployer.declareIfNot({
    contract: PositionsCompiledContract as any,
    casm: PositionsCompiledContractCASM as any,
  });
  const routerDeclare = await deployer.declareIfNot({
    contract: Router as any,
    casm: RouterCASM as any,
  });
  const twammDeclare = await deployer.declareIfNot({
    contract: TWAMMCompiledContract as any,
    casm: TWAMMCompiledContractCASM as any,
  });

  const {
    contract_address: [coreAddress],
  } = await deployer.deploy({
    classHash: coreContractDeclare.class_hash,
    constructorCalldata: [deployer.address],
    salt: "0x0",
  });

  const positionsConstructorCalldata = [
    deployer.address,
    coreAddress,
    declareNftResponse.class_hash,
    shortString.encodeShortString("https://f.ekubo.org/"),
  ];

  const {
    contract_address: [positionsAddress, routerAddress, twammAddress],
  } = await deployer.deploy([
    {
      classHash: positionsDeclare.class_hash,
      constructorCalldata: positionsConstructorCalldata,
      salt: "0x1",
    },
    {
      classHash: routerDeclare.class_hash,
      constructorCalldata: [coreAddress],
      salt: "0x2",
    },
    {
      classHash: twammDeclare.class_hash,
      constructorCalldata: [deployer.address, coreAddress],
      salt: "0x3",
    },
  ]);

  const positions = new Contract(
    PositionsCompiledContract.abi,
    positionsAddress,
    deployer
  );

  await positions.invoke("set_twamm", [twammAddress]);

  const nftAddress = (await positions.call("get_nft_address")) as bigint;

  return {
    core: coreAddress,
    positions: positionsAddress,
    router: routerAddress,
    nft: num.toHexString(nftAddress),
    twamm: twammAddress,
    tokenClassHash: simpleTokenContractDeclare.class_hash,
  };
}

export async function prepareContracts(
  setup: Awaited<ReturnType<typeof setupContracts>>
) {
  const account = await createAccount();
  const getTxSettings = await getNextTransactionSettingsFunction(
    account,
    "0x1"
  );

  const core = new Contract(CoreContract.abi, setup.core, account);
  const nft = new Contract(OwnedNFTContract.abi, setup.nft, account);
  const positionsContract = new Contract(
    PositionsContract.abi,
    setup.positions,
    account
  );
  const router = new Contract(RouterContract.abi, setup.router, account);
  const twamm = new Contract(TWAMMCompiledContract.abi, setup.twamm, account);

  const [token0Address, token1Address] = await deployTokens({
    deployer: account,
    classHash: setup.tokenClassHash,
    getTxSettings,
  });

  const token0 = new Contract(MockERC20Contract.abi, token0Address, account);
  const token1 = new Contract(MockERC20Contract.abi, token1Address, account);

  return {
    account,
    core,
    twamm,
    nft,
    positionsContract,
    router,
    token0,
    token1,
    getTxSettings,
  };
}
