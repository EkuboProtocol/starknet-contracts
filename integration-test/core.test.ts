import { BlockTag, Call } from "starknet";
import { POOL_CASES } from "./cases/poolCases";
import { SWAP_CASES } from "./cases/swapCases";
import { getAmountsForLiquidity } from "./utils/liquidityMath";
import { prepareContracts, setupContracts } from "./utils/setupContracts";
import { fromI129, i129, toI129 } from "./utils/serialize";
import { provider, setDevnetTime } from "./utils/provider";
import { computeFee } from "./utils/computeFee";
import { beforeAll, describe, it } from "vitest";
import { formatPrice } from "./utils/formatPrice";
import { TWAMM_POOL_CASES, TWAMM_SWAP_CASES } from "./cases/twammCases";
import { MAX_BOUNDS_TWAMM, MAX_TICK_SPACING } from "./utils/constants";

describe("core", () => {
  let setup: Awaited<ReturnType<typeof setupContracts>>;

  beforeAll(async () => {
    setup = await setupContracts({
      core: "0x7bebe73b57806307db657aa0bc94a482c8489b9dd5abca1048c9f39828e6907",
      positions:
        "0x22c38145f143fc7bb979c6cac6d76bfb719bbb1e4bba9764ebafd0a7890367",
      router:
        "0x28078e4b34563d7259c79b86378568b30660e30b1e9def79b845a32eed7ea7e",
      nft: "0x512e23b626c051d76fb95b72696fbed63c821e499d90aa523ac523a1246a2c8",
      twamm:
        "0x78bd51ff6ac4d7b0cd598435b49e29a89912f8d9e40c9abc5af0dba1fe133e9",
      tokenClassHash:
        "0x645bbd4bf9fb2bd4ad4dd44a0a97fa36cce3f848ab715ddb82a093337c1e42e",
    });
    console.log(setup);
  }, 300_000);

  describe("regular pool", () => {
    for (const { name: poolCaseName, pool, positions } of POOL_CASES) {
      describe.concurrent(poolCaseName, () => {
        for (const swapCase of SWAP_CASES) {
          it(`swap ${swapCase.amount} ${
            swapCase.isToken1 ? "token1" : "token0"
          }${swapCase.skipAhead ? ` skip ${swapCase.skipAhead}` : ""}${
            swapCase.sqrtRatioLimit
              ? ` limit ${formatPrice(swapCase.sqrtRatioLimit)}`
              : ""
          }`, async ({ expect }) => {
            const {
              nft,
              account,
              token1,
              token0,
              router,
              positionsContract,
              core,
              getTxSettings,
            } = await prepareContracts(setup);

            const poolKey = {
              token0: token0.address,
              token1: token1.address,
              fee: pool.fee,
              tick_spacing: pool.tickSpacing,
              extension: "0x0",
            };

            const txHashes: string[] = [];
            for (const { liquidity, bounds } of positions) {
              const { amount0, amount1 } = getAmountsForLiquidity({
                tick: pool.startingTick,
                liquidity,
                bounds,
              });
              const { transaction_hash } = await account.execute(
                [
                  core.populate("maybe_initialize_pool", [
                    poolKey,

                    // starting tick
                    toI129(pool.startingTick),
                  ]),
                  token0.populate("transfer", [setup.positions, amount0]),
                  token1.populate("transfer", [setup.positions, amount1]),
                  positionsContract.populate(
                    "mint_and_deposit_and_clear_both",
                    [
                      poolKey,
                      {
                        lower: toI129(bounds.lower),
                        upper: toI129(bounds.upper),
                      },
                      liquidity,
                    ]
                  ),
                ],
                [],
                getTxSettings()
              );

              txHashes.push(transaction_hash);
            }

            const mintReceipts = await Promise.all(
              txHashes.map((txHash) =>
                provider.waitForTransaction(txHash, {
                  retryInterval: 0,
                })
              )
            );

            const positionsMinted: { token_id: bigint; liquidity: bigint }[] =
              mintReceipts.map((receipt) => {
                const { Transfer } = nft
                  .parseEvents(receipt)
                  .find(({ Transfer }) => Transfer);

                const { PositionUpdated } = core
                  .parseEvents(receipt)
                  .find(({ PositionUpdated }) => PositionUpdated);

                return {
                  token_id: (Transfer as unknown as { token_id: bigint })
                    .token_id,
                  liquidity: (
                    PositionUpdated as unknown as {
                      params: {
                        liquidity_delta: { mag: bigint; sign: boolean };
                      };
                    }
                  ).params.liquidity_delta.mag,
                };
              });

            const [remaining0, remaining1] = await Promise.all([
              token0.call("balanceOf", [account.address]),
              token1.call("balanceOf", [account.address]),
            ]);
            // transfer remaining balances to router
            await account.execute(
              [
                token0.populate("transfer", [setup.router, remaining0]),
                token1.populate("transfer", [setup.router, remaining1]),
              ],
              [],
              getTxSettings()
            );

            let transaction_hash: string;
            try {
              ({ transaction_hash } = await router.invoke(
                "swap",
                [
                  {
                    pool_key: poolKey,
                    sqrt_ratio_limit: swapCase.sqrtRatioLimit ?? 0,
                    skip_ahead: swapCase.skipAhead ?? 0,
                  },
                  {
                    amount: toI129(swapCase.amount),
                    token: swapCase.isToken1 ? token1.address : token0.address,
                  },
                ],
                getTxSettings()
              ));
            } catch (error) {
              transaction_hash = error.transaction_hash;
              if (!transaction_hash) throw error;
            }

            const swap_receipt = await provider.waitForTransaction(
              transaction_hash,
              { retryInterval: 0 }
            );

            switch (swap_receipt.execution_status) {
              case "REVERTED": {
                const revertReason = swap_receipt.revert_reason;

                const hexErrorMessage =
                  /Failure reason: 0x([a-fA-F0-9]+)/g.exec(revertReason)?.[1];

                expect({
                  revert_reason: hexErrorMessage
                    ? Buffer.from(hexErrorMessage, "hex").toString("ascii")
                    : /(RunResources has no remaining steps)/g.exec(
                        revertReason
                      )?.[1] ?? revertReason,
                }).toMatchSnapshot();
                break;
              }
              case "SUCCEEDED": {
                const execution_resources = swap_receipt.execution_resources;
                if ("memory_holes" in execution_resources) {
                  delete execution_resources["memory_holes"];
                }

                const { sqrt_ratio_after, tick_after, liquidity_after, delta } =
                  core.parseEvents(swap_receipt)[0].Swapped;

                const { amount0, amount1 } = delta as unknown as {
                  amount0: i129;
                  amount1: i129;
                };

                expect({
                  execution_resources,
                  delta: {
                    amount0: fromI129(amount0),
                    amount1: fromI129(amount1),
                  },
                  liquidity_after,
                  sqrt_ratio_after,
                  tick_after: fromI129(tick_after as unknown as i129),
                }).toMatchSnapshot();
                break;
              }
            }

            let cumulativeProtocolFee0 = 0n;
            let cumulativeProtocolFee1 = 0n;

            const withdrawalCalls: Call[] = [];
            for (let i = 0; i < positions.length; i++) {
              const { bounds } = positions[i];

              const boundsArgument = {
                lower: toI129(bounds.lower),
                upper: toI129(bounds.upper),
              };

              const { liquidity: expectedLiquidity, token_id } =
                positionsMinted[i];

              const { liquidity, amount0, amount1 } =
                (await positionsContract.call(
                  "get_token_info",
                  [token_id, poolKey, boundsArgument],
                  { blockIdentifier: BlockTag.pending }
                )) as unknown as {
                  liquidity: bigint;
                  amount0: bigint;
                  amount1: bigint;
                  fees0: bigint;
                  fees1: bigint;
                };

              expect(liquidity).toEqual(expectedLiquidity);

              cumulativeProtocolFee0 += computeFee(amount0, poolKey.fee);
              cumulativeProtocolFee1 += computeFee(amount1, poolKey.fee);

              withdrawalCalls.push(
                positionsContract.populate("withdraw", [
                  token_id,
                  poolKey,
                  boundsArgument,
                  liquidity,
                  0,
                  0,
                  // collect_fees =
                  true,
                ])
              );
            }

            // wait for all the withdrawals to be mined
            await account.execute(withdrawalCalls, [], getTxSettings());

            const [protocolFee0, protocolFee1, balance0, balance1] =
              await Promise.all([
                core.call("get_protocol_fees_collected", [token0.address]),
                core.call("get_protocol_fees_collected", [token1.address]),

                token0.call("balanceOf", [setup.core]),
                token1.call("balanceOf", [setup.core]),
              ]);

            expect(protocolFee0).toEqual(cumulativeProtocolFee0);
            expect(protocolFee1).toEqual(cumulativeProtocolFee1);

            // assuming up to 1 wei of rounding error per swap / withdrawal
            expect(balance0).toBeGreaterThanOrEqual(cumulativeProtocolFee0);
            expect(balance1).toBeGreaterThanOrEqual(cumulativeProtocolFee1);

            // extra is just to account for rounding error for position mints and withdraws as well as swaps (each iteration causes rounding error)
            expect(balance0).toBeLessThanOrEqual(cumulativeProtocolFee0 + 200n);
            expect(balance1).toBeLessThanOrEqual(cumulativeProtocolFee1 + 200n);
          }, 300_000);
        }
      });
    }
  });

  describe.only("twamm", () => {
    for (const {
      name: twammCaseName,
      pool,
      positions_liquidities,
    } of TWAMM_POOL_CASES) {
      describe(twammCaseName, () => {
        for (const { name, orders, snapshot_times } of TWAMM_SWAP_CASES) {
          it(
            name,
            async ({ expect }) => {
              const {
                account,
                core,
                positionsContract,
                token0,
                token1,
                twamm,
                getTxSettings,
              } = await prepareContracts(setup);

              const poolKey = {
                token0: token0.address,
                token1: token1.address,
                fee: pool.fee,
                tick_spacing: MAX_TICK_SPACING,
                extension: twamm.address,
              };

              const stateKey = {
                token0: poolKey.token0,
                token1: poolKey.token1,
                fee: poolKey.fee,
              };

              const txHashes: string[] = [];
              for (const liquidity of positions_liquidities) {
                const bounds = MAX_BOUNDS_TWAMM;
                const { amount0, amount1 } = getAmountsForLiquidity({
                  tick: pool.startingTick,
                  liquidity,
                  bounds,
                });
                const { transaction_hash } = await account.execute(
                  [
                    core.populate("maybe_initialize_pool", [
                      poolKey,

                      // starting tick
                      toI129(pool.startingTick),
                    ]),
                    token0.populate("transfer", [setup.positions, amount0]),
                    token1.populate("transfer", [setup.positions, amount1]),
                    positionsContract.populate(
                      "mint_and_deposit_and_clear_both",
                      [
                        poolKey,
                        {
                          lower: toI129(bounds.lower),
                          upper: toI129(bounds.upper),
                        },
                        liquidity,
                      ]
                    ),
                  ],
                  [],
                  getTxSettings()
                );

                txHashes.push(transaction_hash);
              }

              await Promise.all(
                txHashes.map((txHash) =>
                  provider.waitForTransaction(txHash, {
                    retryInterval: 0,
                  })
                )
              );

              const startingTime = 16; // (await provider.getBlock("pending")).timestamp;
              await setDevnetTime(startingTime - 8);

              console.log(
                "time",
                (await provider.getBlock("pending")).timestamp
              );

              if (orders.length > 0) {
                const [balance0, balance1] = await Promise.all([
                  token0.balanceOf(account.address),
                  token1.balanceOf(account.address),
                ]);

                const { transaction_hash } = await account.execute(
                  [
                    token0.populate("transfer", [setup.positions, balance0]),
                    token1.populate("transfer", [setup.positions, balance1]),
                    ...orders.map((order) => {
                      const [buy_token, sell_token] = order.is_token1
                        ? [poolKey.token0, poolKey.token1]
                        : [poolKey.token1, poolKey.token0];

                      return positionsContract.populate(
                        "mint_and_increase_amount",
                        [
                          {
                            sell_token,
                            buy_token,
                            fee: poolKey.fee,
                            start_time:
                              startingTime + order.relative_times.start,
                            end_time: startingTime + order.relative_times.end,
                          },
                          order.amount,
                        ]
                      );
                    }),
                  ],
                  [],
                  getTxSettings()
                );

                const orderPlacementReceipt = await account.waitForTransaction(
                  transaction_hash,
                  { retryInterval: 0 }
                );

                expect(
                  orderPlacementReceipt.execution_status,
                  "order placement succeeded"
                ).toEqual("SUCCEEDED");

                const orderUpdatedEvents = twamm
                  .parseEvents(orderPlacementReceipt)
                  .map(({ OrderUpdated }) => OrderUpdated)
                  .filter((x) => !!x);

                console.log("order updated events", orderUpdatedEvents);
              }

              console.log(
                "#get_last_virtual_order_time",
                await twamm.get_last_virtual_order_time(stateKey)
              );

              for (const snapshotTime of snapshot_times) {
                await setDevnetTime(startingTime + snapshotTime);

                const { transaction_hash } = await account.execute(
                  [
                    twamm.populate("execute_virtual_orders", [
                      {
                        token0: poolKey.token0,
                        token1: poolKey.token1,
                        fee: poolKey.fee,
                      },
                    ]),
                  ],
                  [],
                  getTxSettings()
                );

                const executeVirtualOrdersReceipt =
                  await account.waitForTransaction(transaction_hash, {
                    retryInterval: 0,
                  });

                console.log(executeVirtualOrdersReceipt);

                console.log(
                  "#get_last_virtual_order_time",
                  await twamm.get_last_virtual_order_time(stateKey)
                );

                const twammEvents = twamm.parseEvents(
                  executeVirtualOrdersReceipt
                );
                const coreEvents = core.parseEvents(
                  executeVirtualOrdersReceipt
                );
                expect({ twammEvents, coreEvents }).toMatchSnapshot();
              }
            },
            300_000
          );
        }
      });
    }
  });
});
