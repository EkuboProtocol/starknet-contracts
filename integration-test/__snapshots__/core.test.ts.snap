// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core tests 1e18 full range liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap -10000 token0 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 27,
      "pedersen_builtin": 79,
      "range_check_builtin": 1879,
    },
    "n_memory_holes": 7841,
    "n_steps": 25830,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 10000n,
        "sign": true,
      },
      "amount1": {
        "mag": 10030n,
        "sign": false,
      },
    },
    "liquidity_after": 1000000000000000000n,
    "sqrt_ratio_after": 340282366920941876495514824478789461470n,
    "tick_after": {
      "mag": 0n,
      "sign": false,
    },
  },
}
`;

exports[`core tests 1e18 full range liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap -10000 token1 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 81,
      "pedersen_builtin": 85,
      "range_check_builtin": 1916,
    },
    "n_memory_holes": 17862,
    "n_steps": 27895,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 10030n,
        "sign": false,
      },
      "amount1": {
        "mag": 10000n,
        "sign": true,
      },
    },
    "liquidity_after": 1000000000000000000n,
    "sqrt_ratio_after": 340282366920935050431234390418979673808n,
    "tick_after": {
      "mag": 1n,
      "sign": true,
    },
  },
}
`;

exports[`core tests 1e18 full range liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap 10000 token0 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 81,
      "pedersen_builtin": 85,
      "range_check_builtin": 1854,
    },
    "n_memory_holes": 18159,
    "n_steps": 27534,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 10000n,
        "sign": false,
      },
      "amount1": {
        "mag": 9969n,
        "sign": true,
      },
    },
    "liquidity_after": 1000000000000000000n,
    "sqrt_ratio_after": 340282366920935070848176405709111855138n,
    "tick_after": {
      "mag": 1n,
      "sign": true,
    },
  },
}
`;

exports[`core tests 1e18 full range liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap 10000 token1 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 27,
      "pedersen_builtin": 79,
      "range_check_builtin": 1830,
    },
    "n_memory_holes": 8043,
    "n_steps": 25559,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 9969n,
        "sign": true,
      },
      "amount1": {
        "mag": 10000n,
        "sign": false,
      },
    },
    "liquidity_after": 1000000000000000000n,
    "sqrt_ratio_after": 340282366920941856078572809188248941300n,
    "tick_after": {
      "mag": 0n,
      "sign": false,
    },
  },
}
`;

exports[`core tests no liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap -10000 token0 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 1620,
      "pedersen_builtin": 402,
      "range_check_builtin": 27633,
    },
    "n_memory_holes": 318074,
    "n_steps": 220022,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 0n,
        "sign": false,
      },
      "amount1": {
        "mag": 0n,
        "sign": false,
      },
    },
    "liquidity_after": 0n,
    "sqrt_ratio_after": 6277100250585753475930931601400621808602321654880405518632n,
    "tick_after": {
      "mag": 88722883n,
      "sign": false,
    },
  },
}
`;

exports[`core tests no liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap -10000 token1 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 1647,
      "pedersen_builtin": 408,
      "range_check_builtin": 27038,
    },
    "n_memory_holes": 327433,
    "n_steps": 223179,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 0n,
        "sign": false,
      },
      "amount1": {
        "mag": 0n,
        "sign": false,
      },
    },
    "liquidity_after": 0n,
    "sqrt_ratio_after": 18446748437148339061n,
    "tick_after": {
      "mag": 88722884n,
      "sign": true,
    },
  },
}
`;

exports[`core tests no liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap 10000 token0 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 1647,
      "pedersen_builtin": 408,
      "range_check_builtin": 26978,
    },
    "n_memory_holes": 327733,
    "n_steps": 222707,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 0n,
        "sign": false,
      },
      "amount1": {
        "mag": 0n,
        "sign": true,
      },
    },
    "liquidity_after": 0n,
    "sqrt_ratio_after": 18446748437148339061n,
    "tick_after": {
      "mag": 88722884n,
      "sign": true,
    },
  },
}
`;

exports[`core tests no liquidity, starting at price 1, tick_spacing=0.6%, fee=0.3% swap 10000 token1 1`] = `
{
  "execution_resources": {
    "builtin_instance_counter": {
      "bitwise_builtin": 1620,
      "pedersen_builtin": 402,
      "range_check_builtin": 27574,
    },
    "n_memory_holes": 318361,
    "n_steps": 219550,
  },
  "result": {
    "delta": {
      "amount0": {
        "mag": 0n,
        "sign": true,
      },
      "amount1": {
        "mag": 0n,
        "sign": false,
      },
    },
    "liquidity_after": 0n,
    "sqrt_ratio_after": 6277100250585753475930931601400621808602321654880405518632n,
    "tick_after": {
      "mag": 88722883n,
      "sign": false,
    },
  },
}
`;
